pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -1072,6 +1093,82 @@ func (nc *Controller) handleDisruption(zoneToNodeConditions map[string][]*v1.Nod
 	}
 }
 
+func (nc *Controller) podUpdated(oldPod, newPod *v1.Pod) {
+	if len(newPod.Spec.NodeName) != 0 && (oldPod == nil || newPod.Spec.NodeName != oldPod.Spec.NodeName) {
+		nc.podUpdateQueue.Add(newPod)
+	}
+}
+
+func (nc *Controller) doPodProcessingWorker() {
+	for {
+		obj, shutdown := nc.podUpdateQueue.Get()
+		// "podUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
+			return
+		}
+
+		pod := obj.(*v1.Pod)
+		nodeName := pod.Spec.NodeName
+
+		func() {
+			nc.nodeHealthLock.Lock()
+			defer nc.nodeHealthLock.Unlock()
+			defer nc.podUpdateQueue.Done(pod)
+
+			nodeHealth, exists := nc.nodeHealthMap[nodeName]
+			if !exists {
+				return
+			}
+
+			node, err := nc.nodeLister.Get(nodeName)
+			if err != nil {
+				klog.Errorf("Failed to list node %v.", nodeName)
+				return
+			}
+
+			gracePeriod := nc.nodeMonitorGracePeriod
+			_, currentReadyCondition := nodeutil.GetNodeCondition(nodeHealth.status, v1.NodeReady)
+			if currentReadyCondition == nil {
+				// Nothing to do. Pod will be handled by monitorNodeHealth when needed.
+				return
+			}
+
+			decisionTimestamp := nc.now()
+			if nc.useTaintBasedEvictions {
+				if currentReadyCondition.Status == v1.ConditionFalse &&
+					decisionTimestamp.After(nodeHealth.readyTransitionTimestamp.Add(nc.podEvictionTimeout)) {
+					if nc.evictPods(node) {
```

Comment:
Moreover, this isn't really behaving how we would like this to behave.

evictPods is actually only marking a node to evict all the pods that are scheduled on that node.
So if we will call evictPods() for the same node multiple times, if our cache for pods is lagging, we will be trying to evict the same pods over and over again.

Unless I'm not missing anything, I think the logic should be main like: "evictPods(node, pods)" (that will only be appending given pods for a particular node for eviction).